ide/coqide/idetop.ml

```diff
@@ -88,14 +88,21 @@ let ide_doc = ref None
 let get_doc () = Option.get !ide_doc
 let set_doc doc = ide_doc := Some doc
 
+(* This patches up the offset to compensate the XML protocol not
+   having a notion of absolute offset *)
+let fix_loc ~off loc =
+  Loc.{ loc with bol_pos = loc.bol_pos + off;
+                 bp = loc.bp + off;
+                 ep = loc.ep + off}
```

Comment:
Actually I have no idea why fix_loc was necessary in the first place, and it created a serious bug in the first place (https://github.com/coq/coq/issues/15087) . This patch just moves fix_loc to be coqide specific.

Reply:
> Actually I have no idea why fix_loc was necessary in the first place, and it created a serious bug (https://github.com/coq/coq/issues/15087).

`fix_loc` is "necessary" because #16978 didn't fix #15087.

Here's an excerpt from `Idetop.add`, lines 103-105.  The first line includes `bp=off` to initialize the parser, **which is ignored**.  That's the bug of #15087.  Adding a printf in `fix_loc` shows that bp/ep are relative to the beginning of the statement rather than the beginning of the buffer ("absolute").

What you have in this PR works for CoqIDE, but it seems likely you will need a similar `fix_loc` to support the debugger in LSP, too--at least if you want breakpoints that are after the first sentence to work.  Better to fix this in the lexer/parser than in both XML and LSP.

You may want to revert the `fix_loc` changes in this PR.  Fixing the lexer/parser can be done later.  I won't object if you and @Zimmi48 want to merge this as-is.

```
  let loc = { (initial ToplevelInput) with bp=off; line_nb } in
  let fix_loc = fix_loc ~off in
  let pa = Pcoq.Parsable.make ~loc ~fix_loc (Gramlib.Stream.of_string s) in
```

```
bp = 4 ep = 8 line_nb = 5 off = 119
bp = 9 ep = 13 line_nb = 5 off = 119
bp = 13 ep = 14 line_nb = 5 off = 119
bp = 2 ep = 7 line_nb = 6 off = 133
bp = 7 ep = 8 line_nb = 6 off = 133
bp = 2 ep = 6 line_nb = 7 off = 141
bp = 6 ep = 7 line_nb = 7 off = 141
bp = 11 ep = 14 line_nb = 8 off = 148
bp = 14 ep = 15 line_nb = 8 off = 148
bp = 4 ep = 8 line_nb = 10 off = 163
bp = 9 ep = 13 line_nb = 10 off = 163
```

Reply:
This is not how Coq's lexer and parser work, bp will be set from `Stream.count ()`, so in the above case you need to properly call `Stream.of_string ~offset`  with the right offset. Please read #16978 again.

The above code is indeed nonsensical but equivalent to what was in 8.16, the goal is to preserve the exact behavior as the hack removed in the fix set some parts of the location at random.

There is no fixing to do w.r.t. parsing resumption (which has been working since a long time in LSP).

Reply:
Then there's no point in creating `fix_loc` at all.  The offset should be passed in the in the call setting up the stream.  The code should be like this:

```
let add ((((s,eid),(sid,verbose)),offset),(line_nb,bol_pos)) =
  let doc = get_doc () in
  (* note: this won't yield correct values for bol_pos_last,
     but the debugger doesn't use that *)
  let loc = Loc.{ (initial ToplevelInput) with line_nb } in
  let pa = Pcoq.Parsable.make ~loc (Gramlib.Stream.of_string ~offset s) in
```

Of course, the `offset` option didn't exist when I was working on this.

Reply:
This is what master does.